apps/wallet/src/ui/app/helpers/getEventsSummary.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { notEmpty } from '_helpers';
+
+import type { TransactionEffects } from '@mysten/sui.js';
+
+export type CoinsMetaProps = {
+    amount: number;
+    coinType: string;
+    receiverAddress: string;
+};
+
+export type TxnMetaResponse = {
+    objectIDs: string[];
+    coins: CoinsMetaProps[];
+};
+
+export function getEventsSummary(
+    txEffects: TransactionEffects,
+    address: string
+): TxnMetaResponse {
+    const events = txEffects?.events || [];
+
+    const coinsMeta = events
+        .map((event) => {
+            if (
+                'coinBalanceChange' in event &&
+                ['Receive', 'Pay'].includes(
+                    event?.coinBalanceChange?.changeType
+                )
+            ) {
+                /// A net positive amount means the user received coins
+                /// A net negative amount means the user sent coins
+                const { coinBalanceChange } = event;
+                const { coinType, amount, coinObjectId, owner } =
+                    coinBalanceChange;
+                const { AddressOwner } = owner as { AddressOwner: string };
+                const { ObjectOwner } = owner as { ObjectOwner: string };
+
+                if (ObjectOwner) {
+                    // TODO - update once the issue with the ObjectOwner is fixed
+                    return null;
+                }
+
+                return {
+                    amount: amount,
+                    coinType: coinType,
+                    coinObjectId: coinObjectId,
+                    receiverAddress: AddressOwner,
+                };
+            }
+            return null;
+        })
+        .filter(notEmpty);
+    const objectIDs: string[] = events
+
+        .map((event) => {
+            if (!('transferObject' in event)) {
+                return null;
+            }
+            const { transferObject } = event;
+            const { AddressOwner } = transferObject.recipient as {
+                AddressOwner: string;
+            };
+            if (AddressOwner !== address) {
+                return null;
+            }
+            return transferObject?.objectId;
+        })
+        .filter(notEmpty);
+
+    /// Group coins by receiverAddress
+    // sum coins by coinType for each receiverAddress
+    const meta = coinsMeta.reduce((acc, value, _) => {
+        return {
```

Comment:
I wonder if this would be cleaner to just do this when you construct `coinsMeta` instead. Instead of a map just make it a `forEach` where you insert into an existing object. Seems like it'd simplify things.

Reply:
Done